File: trust-tasks-dart-didcomm/lib/src/envelope.dart
import 'dart:convert';
import 'dart:math';

import 'package:didcomm/didcomm.dart';
import 'package:ssi/ssi.dart';
import 'package:trust_tasks/trust_tasks.dart';

/// The binding's stable identifier (`bindings/didcomm/0.2` §1).
const String bindingUri = _bindingUri;

// Named privately as well, because [DidcommTransport.bindingUri] shadows the
// public constant inside that class.
const String _bindingUri = 'https://trusttasks.org/binding/didcomm/0.2';

/// The DIDComm message `type` that carries a Trust Task document (§1). It did
/// not change with the binding's `0.1` → `0.2` minor; only the binding
/// identifier did.
const String envelopeType =
    'https://trusttasks.org/binding/didcomm/0.1/envelope';

/// The wrappings that authenticate a sender: authcrypt, optionally signed
/// inside, optionally anoncrypted outside for a mediator. Anoncrypt alone,
/// signed-only and plaintext are refused (§2, §4).
const List<MessageWrappingType> authenticatedWrappings = [
  MessageWrappingType.authcryptPlaintext,
  MessageWrappingType.authcryptSignPlaintext,
  MessageWrappingType.anoncryptAuthcryptPlaintext,
];

/// The [TransportHandler] for one DIDComm exchange: [peer] is the DID the
/// authcrypt authenticated, [local] the DID that unpacked it (§3).
final class DidcommTransport implements TransportHandler {
  const DidcommTransport({this.local, this.peer});

  final String? local;
  final String? peer;

  @override
  String get bindingUri => _bindingUri;

  @override
  TransportContext deriveParties() =>
      TransportContext(issuer: peer, recipient: local);
}

/// Why an envelope was refused before a document could be produced.
///
/// Except for [invalidBody], none of these reach the framework pipeline, and
/// none can be answered with a `trust-task-error`: there is no authenticated
/// sender to address one to (§4, `errored`).
enum EnvelopeFailure {
  /// Not a DIDComm message this agent can open: not JSON, not a JWE, or
  /// decryption failed.
  undecryptable,

  /// Anoncrypt, signed-only or plaintext: no authenticated sender (§4).
  unauthenticatedSender,

  /// The authcrypt `skid` carries no `#fragment`, so it names no key.
  unqualifiedSenderKid,

  /// The sender is not on the allowlist. Refused before decryption, so an
  /// unlisted sender's DID is never resolved.
  senderNotAllowed,

  /// The DIDComm `type` is not [envelopeType] (§2 item 2).
  wrongEnvelopeType,

  /// The body is not a Trust Task document. The sender *is* authenticated here,
  /// so [EnvelopeException.sender] is set and a caller may answer with
  /// `malformedRequest` (§4).
  invalidBody,
}

/// An envelope refused by [unpackTrustTask].
final class EnvelopeException implements Exception {
  const EnvelopeException(this.failure, this.detail, {this.sender});

  final EnvelopeFailure failure;

  /// For logs. It can name DIDs; do not send it to the peer.
  final String detail;

  /// The authenticated sender, where authentication got that far.
  final String? sender;

  @override
  String toString() => 'EnvelopeException($failure): $detail';
}

/// A Trust Task document taken out of an authcrypt envelope.
final class UnpackedTrustTask {
  const UnpackedTrustTask({
    required this.document,
    required this.transport,
    required this.message,
  });

  /// The Trust Task document, as JSON — exactly as the sender serialised it,
  /// which is what a proof over it was computed on.
  final Map<String, dynamic> document;

  /// The authenticated sender and the unpacking recipient.
  final DidcommTransport transport;

  /// The DIDComm message that carried it, for its headers.
  final PlainTextMessage message;
}

/// Pack [document] into an authcrypt envelope from [sender] to [recipient]
/// (§2), returning the JWE as JSON.
///
/// `thid` is set from the document's `threadId`, falling back to its `id`, and
/// `pthid` from `parentThreadId` — the headers follow the members, never the
/// reverse (§3.1). The sender's key-agreement key is the first one whose curve
/// the recipient also has.
Future<Map<String, dynamic>> packTrustTask(
  Map<String, dynamic> document, {
  required DidManager sender,
  required DidDocument recipient,
}) async {
  final senderDoc = await sender.getDidDocument();
  final keyIds =
      senderDoc.matchKeysInKeyAgreement(otherDidDocuments: [recipient]);
  if (keyIds.isEmpty) {
    throw ArgumentError.value(
      recipient.id,
      'recipient',
      'shares no key-agreement curve with ${senderDoc.id}',
    );
  }
  final id = document['id'];
  final threadId = document['threadId'] ?? id;
  final parentThreadId = document['parentThreadId'];

  final message = PlainTextMessage(
    id: _uuid(),
    type: Uri.parse(envelopeType),
    from: senderDoc.id,
    to: [recipient.id],
    threadId: threadId is String ? threadId : null,
    parentThreadId: parentThreadId is String ? parentThreadId : null,
    createdTime: DateTime.now().toUtc(),
    body: document,
  );

  final encrypted = await EncryptedMessage.packWithAuthentication(
    message,
    keyPair: await sender.getKeyPairByDidKeyId(keyIds.first),
    didKeyId: keyIds.first,
    recipientDidDocuments: [recipient],
  );
  return jsonDecode(jsonEncode(encrypted)) as Map<String, dynamic>;
}

/// The sender DID an authcrypt envelope *claims*, read from its protected
/// header without decrypting anything. Unauthenticated until the envelope is
/// opened — use it only to decide whether to open it.
String? advertisedSender(Map<String, dynamic> envelope) {
  try {
    final header = jsonDecode(
      utf8.decode(
        base64Url.decode(base64Url.normalize(envelope['protected'] as String)),
      ),
    ) as Map<String, dynamic>;
    final skid = header['skid'];
    if (skid is! String || !skid.contains('#')) return null;
    return skid.substring(0, skid.indexOf('#'));
  } on Object {
    return null;
  }
}

/// Open an authcrypt envelope addressed to [recipient] and take out the Trust
/// Task document (§2 consumer items 1–3, and §3.1's thread check is left to
/// the pipeline, which can answer it).
///
/// With [allowedSenders], an envelope whose `skid` names anyone else is refused
/// **before** decryption: opening it means resolving the sender's DID, which
/// for `did:web` is a fetch from a host the sender picked.
///
/// Throws [EnvelopeException].
Future<UnpackedTrustTask> unpackTrustTask(
  Map<String, dynamic> envelope, {
  required DidManager recipient,
  Set<String>? allowedSenders,
}) async {
  if (allowedSenders != null) {
    final claimed = advertisedSender(envelope);
    if (claimed == null) {
      throw const EnvelopeException(
        EnvelopeFailure.unauthenticatedSender,
        'no authcrypt skid to check against the allowlist',
      );
    }
    if (!allowedSenders.contains(claimed)) {
      throw EnvelopeException(
        EnvelopeFailure.senderNotAllowed,
        'sender $claimed is not on this consumer\'s allowlist',
      );
    }
  }

  List<DidcommMessage> layers = const [];
  final PlainTextMessage message;
  try {
    // Every wrapping is let through here and judged below from the layers
    // actually found, rather than by reading didcomm's error text.
    message = await DidcommMessage.unpackToPlainTextMessage(
      message: envelope,
      recipientDidManager: recipient,
      expectedMessageWrappingTypes: MessageWrappingType.values,
      onUnpacked: ({required foundMessages, required foundSigners}) {
        layers = foundMessages;
      },
    );
  } on Object catch (e) {
    // Not a DIDComm message, not for this recipient, a failed key unwrap or
    // AEAD tag, or a `from` / `to` that disagrees with the JWE's own headers.
    throw EnvelopeException(EnvelopeFailure.undecryptable, '$e');
  }

  final wrapping = MessageWrappingType.findFromMessages(layers);
  if (wrapping == null || !authenticatedWrappings.contains(wrapping)) {
    throw EnvelopeException(
      EnvelopeFailure.unauthenticatedSender,
      'wrapping ${wrapping?.name ?? 'unrecognised'} authenticates no sender',
    );
  }

  // The sender is the skid of the ECDH-1PU layer: the key whose public half
  // opened the key wrap. didcomm has already checked the message's `from`
  // against it.
  String? skid;
  for (final layer in layers) {
    if (layer is! EncryptedMessage) continue;
    final header = _protectedHeader(layer.protected);
    if (header?['alg'] is String &&
        (header!['alg'] as String).startsWith('ECDH-1PU')) {
      skid = header['skid'] as String?;
    }
  }
  if (skid == null) {
    throw const EnvelopeException(
      EnvelopeFailure.unauthenticatedSender,
      'no ECDH-1PU layer',
    );
  }
  final hash = skid.indexOf('#');
  if (hash <= 0) {
    // Never downgrade to "no sender": the pipeline would then trust the
    // in-band issuer with the transport cross-check skipped.
    throw EnvelopeException(
      EnvelopeFailure.unqualifiedSenderKid,
      'sender kid $skid carries no verification-method fragment',
    );
  }
  final sender = skid.substring(0, hash);
  if (allowedSenders != null && !allowedSenders.contains(sender)) {
    throw EnvelopeException(
      EnvelopeFailure.senderNotAllowed,
      'authenticated sender $sender is not on this consumer\'s allowlist',
    );
  }

  if (message.type.toString() != envelopeType) {
    throw EnvelopeException(
      EnvelopeFailure.wrongEnvelopeType,
      'DIDComm type ${message.type}',
      sender: sender,
    );
  }
  final body = message.body;
  if (body == null || body['id'] is! String || body['type'] is! String) {
    throw EnvelopeException(
      EnvelopeFailure.invalidBody,
      'body is not a Trust Task document',
      sender: sender,
    );
  }

  final local = (await recipient.getDidDocument()).id;
  return UnpackedTrustTask(
    document: body,
    transport: DidcommTransport(local: local, peer: sender),
    message: message,
  );
}

/// §3.1: where a DIDComm thread header and its framework member are *both*
/// present they must agree. Returns the refusal, or null.
///
/// Scoped to both-present because DIDComm's `thid` defaults to the message `id`
/// and the framework's `threadId` to the document `id` — different identifier
/// spaces, so an unconditional comparison would refuse conforming exchanges.
RejectReason? checkThreadHeaders(
  PlainTextMessage message,
  Map<String, dynamic> document,
) {
  for (final (header, member, transport) in [
    ('thid', 'threadId', message.threadId),
    ('pthid', 'parentThreadId', message.parentThreadId),
  ]) {
    final inBand = document[member];
    if (transport != null && inBand is String && transport != inBand) {
      return RejectReason(
        code: StandardCode.malformedRequest,
        message: 'DIDComm $header disagrees with the document\'s $member '
            '(bindings/didcomm/0.2 §3.1)',
      );
    }
  }
  return null;
}

Map<String, dynamic>? _protectedHeader(String protected) {
  try {
    return jsonDecode(
            utf8.decode(base64Url.decode(base64Url.normalize(protected))))
        as Map<String, dynamic>;
  } on Object {
    return null;
  }
}

final Random _random = Random.secure();

String _uuid() {
  final b = List<int>.generate(16, (_) => _random.nextInt(256));
  b[6] = (b[6] & 0x0f) | 0x40;
  b[8] = (b[8] & 0x3f) | 0x80;
  String hex(int from, int to) => [
        for (var i = from; i < to; i++) b[i].toRadixString(16).padLeft(2, '0'),
      ].join();
  return '${hex(0, 4)}-${hex(4, 6)}-${hex(6, 8)}-${hex(8, 10)}-${hex(10, 16)}';
}

/// A fresh `urn:uuid:` identifier, for response and error documents.
String newUrnUuid() => 'urn:uuid:${_uuid()}';

File: trust-tasks-go/didcomm/envelope.go
package didcomm

import (
	"crypto/ecdh"
	"crypto/rand"
	"encoding/json"
	"fmt"
	"strings"

	tt "github.com/trustoverip/dtgwg-trust-tasks-tf/trust-tasks-go/trusttasks"
)

// BindingURI is the binding's stable identifier (bindings/didcomm/0.2 §1). It
// is never on the wire. The identifier moved with the minor; the envelope
// [EnvelopeType] deliberately did not, so a 0.1 and a 0.2 implementation stay
// mutually intelligible on the wire (binding §7.1).
const BindingURI = "https://trusttasks.org/binding/didcomm/0.2"

// EnvelopeType is the DIDComm message `type` a Trust Task envelope carries
// (binding §2). A consumer rejects any other `type`.
const EnvelopeType = "https://trusttasks.org/binding/didcomm/0.1/envelope"

// mediaTypeEncrypted is the JWE `typ` header for a DIDComm v2 encrypted message.
const mediaTypeEncrypted = "application/didcomm-encrypted+json"

// Transport is the [tt.TransportHandler] for one DIDComm v2 exchange: Peer is
// the authcrypt-verified sender DID, Local the DID the message was sealed to.
type Transport struct {
	Local string
	Peer  string
}

// BindingURI implements [tt.TransportHandler].
func (Transport) BindingURI() string { return BindingURI }

// DeriveParties implements [tt.TransportHandler]: the verified peer is the
// issuer, the local DID the recipient; the core applies SPEC §4.8.1 precedence.
func (t Transport) DeriveParties() tt.TransportContext {
	ctx := tt.TransportContext{}
	if t.Peer != "" {
		ctx.Issuer = &t.Peer
	}
	if t.Local != "" {
		ctx.Recipient = &t.Local
	}
	return ctx
}

// Failure is why a DIDComm envelope was refused before a document reached the
// §7.2 pipeline. Binding §4 keeps the first four out of the pipeline entirely —
// there is no authenticated party to route a trust-task-error to.
type Failure int

const (
	// FailNotAuthcrypt: the envelope is plaintext, anoncrypt, or otherwise not
	// an authcrypt JWE. No authenticated sender (binding §2, §4).
	FailNotAuthcrypt Failure = iota
	// FailNotForThisRecipient: no recipient entry this identity can open.
	FailNotForThisRecipient
	// FailSenderUnresolved: the sender's key could not be resolved.
	FailSenderUnresolved
	// FailDecrypt: key unwrap or content decryption failed — a forged skid,
	// wrong recipient, or tampering.
	FailDecrypt
	// FailWrongEnvelopeType: the DIDComm `type` is not [EnvelopeType].
	FailWrongEnvelopeType
	// FailInvalidBody: the message body is not a JSON object / Trust Task shape.
	FailInvalidBody
)

func (f Failure) String() string {
	switch f {
	case FailNotAuthcrypt:
		return "notAuthcrypt"
	case FailNotForThisRecipient:
		return "notForThisRecipient"
	case FailSenderUnresolved:
		return "senderUnresolved"
	case FailDecrypt:
		return "decrypt"
	case FailWrongEnvelopeType:
		return "wrongEnvelopeType"
	case FailInvalidBody:
		return "invalidBody"
	default:
		return "unknown"
	}
}

// EnvelopeError is a DIDComm message refused by [UnpackTrustTask].
type EnvelopeError struct {
	Failure Failure
	Detail  string
	// Sender is the authenticated sender DID URL where authentication got that
	// far; empty otherwise.
	Sender string
}

func (e *EnvelopeError) Error() string {
	return fmt.Sprintf("didcomm: %s: %s", e.Failure, e.Detail)
}

// Identity is a local X25519 key-agreement identity: the DID it belongs to, the
// DID URL (kid) that names the key, and the private key itself.
type Identity struct {
	// DID is the bare DID, e.g. did:example:alice.
	DID string
	// Kid is the key-agreement DID URL, e.g. did:example:alice#key-agreement-1.
	Kid  string
	priv *ecdh.PrivateKey
}

// NewIdentity builds an [Identity] from a 32-byte X25519 private key.
func NewIdentity(did, kid string, x25519Priv []byte) (*Identity, error) {
	priv, err := ecdh.X25519().NewPrivateKey(x25519Priv)
	if err != nil {
		return nil, fmt.Errorf("didcomm: invalid X25519 private key: %w", err)
	}
	return &Identity{DID: did, Kid: kid, priv: priv}, nil
}

// PeerKey is a remote party's X25519 key-agreement public key and the DID URL
// that names it.
type PeerKey struct {
	Kid string
	pub *ecdh.PublicKey
}

// NewPeerKey builds a [PeerKey] from a 32-byte X25519 public key.
func NewPeerKey(kid string, x25519Pub []byte) (*PeerKey, error) {
	pub, err := ecdh.X25519().NewPublicKey(x25519Pub)
	if err != nil {
		return nil, fmt.Errorf("didcomm: invalid X25519 public key: %w", err)
	}
	return &PeerKey{Kid: kid, pub: pub}, nil
}

// didcommMessage is the DIDComm v2 plaintext message (binding §2). Only the
// members the binding populates or reads are modelled; a producer MAY add more,
// and this struct preserves none it does not name.
type didcommMessage struct {
	ID    string          `json:"id"`
	Type  string          `json:"type"`
	From  string          `json:"from,omitempty"`
	To    []string        `json:"to,omitempty"`
	Thid  string          `json:"thid,omitempty"`
	Pthid string          `json:"pthid,omitempty"`
	Body  json.RawMessage `json:"body"`
}

// threadHeaders is the subset of a Trust Task document the binding cross-checks
// against the DIDComm thread headers (binding §3.1).
type threadHeaders struct {
	ID             string `json:"id"`
	ThreadID       string `json:"threadId"`
	ParentThreadID string `json:"parentThreadId"`
}

// Unpacked is a Trust Task document taken out of a DIDComm envelope.
type Unpacked struct {
	// Document is the body, exactly as the sender serialised it.
	Document json.RawMessage
	// Transport is the authenticated sender and opening recipient.
	Transport Transport
	// Thid and Pthid are the DIDComm thread headers, for the §3.1 cross-check
	// the consumer performs against the document's threadId / parentThreadId.
	Thid  string
	Pthid string
}

// PackTrustTask seals doc into a DIDComm v2.1 authcrypt envelope from sender to
// recipient (binding §2, §3.1). doc is the serialised Trust Task document; its
// threadId (or, failing that, its id) becomes the DIDComm `thid`, and its
// parentThreadId becomes `pthid`.
func PackTrustTask(doc json.RawMessage, sender *Identity, recipient *PeerKey) ([]byte, error) {
	var th threadHeaders
	if err := json.Unmarshal(doc, &th); err != nil {
		return nil, fmt.Errorf("didcomm: document is not a JSON object: %w", err)
	}
	thid := th.ThreadID
	if thid == "" {
		thid = th.ID
	}
	msg := didcommMessage{
		ID:    NewMessageID(),
		Type:  EnvelopeType,
		From:  sender.DID,
		To:    []string{bareDID(recipient.Kid)},
		Thid:  thid,
		Pthid: th.ParentThreadID,
		Body:  doc,
	}
	plaintext, err := json.Marshal(msg)
	if err != nil {
		return nil, err
	}
	return packAuthcrypt(plaintext, sender.priv, sender.Kid, recipient.pub, recipient.Kid)
}

// ResolveSender maps an authenticated sender DID to its X25519 key-agreement
// key. The binding does no DID resolution itself (the Go-style model): the
// caller supplies this, looking the DID up in its own store.
type ResolveSender func(senderDID string) (*PeerKey, error)

// UnpackTrustTask opens a DIDComm v2.1 authcrypt envelope addressed to
// recipient and takes out the Trust Task document. The sender's key is resolved
// from the envelope's authenticated skid via resolve. Returns an
// [*EnvelopeError] for a message that never reaches the pipeline (binding §4):
// a non-authcrypt envelope, a wrong `type`, a decryption failure, or an
// unresolvable sender.
func UnpackTrustTask(wire []byte, recipient *Identity, resolve ResolveSender) (*Unpacked, error) {
	inner, err := unpackAuthcrypt(wire, recipient.priv, recipient.Kid, func(skid string) (*ecdh.PublicKey, error) {
		pk, err := resolve(bareDID(skid))
		if err != nil {
			return nil, err
		}
		return pk.pub, nil
	})
	if err != nil {
		return nil, err
	}

	var msg didcommMessage
	if err := json.Unmarshal(inner.plaintext, &msg); err != nil {
		return nil, &EnvelopeError{Failure: FailInvalidBody, Detail: "decrypted payload is not a DIDComm message", Sender: inner.senderKid}
	}
	if msg.Type != EnvelopeType {
		return nil, &EnvelopeError{Failure: FailWrongEnvelopeType, Detail: msg.Type, Sender: inner.senderKid}
	}
	if len(msg.Body) == 0 {
		return nil, &EnvelopeError{Failure: FailInvalidBody, Detail: "message carries no body", Sender: inner.senderKid}
	}

	return &Unpacked{
		Document:  msg.Body,
		Transport: Transport{Local: bareDID(inner.recipientKid), Peer: bareDID(inner.senderKid)},
		Thid:      msg.Thid,
		Pthid:     msg.Pthid,
	}, nil
}

// AdvertisedSender reads the sender DID a JWE envelope's protected header names,
// without opening it — unauthenticated until [UnpackTrustTask] verifies it. Use
// it only to decide whether to open a message, or to resolve the sender's key.
func AdvertisedSender(wire []byte) string {
	var env jweJSON
	if err := json.Unmarshal(wire, &env); err != nil {
		return ""
	}
	protectedJSON, err := b64.DecodeString(env.Protected)
	if err != nil {
		return ""
	}
	var header protectedHeader
	if err := json.Unmarshal(protectedJSON, &header); err != nil {
		return ""
	}
	return bareDID(header.Skid)
}

// bareDID strips the #fragment from a DID URL: did:example:a#key-1 → did:example:a.
func bareDID(didURL string) string {
	if i := strings.IndexByte(didURL, '#'); i >= 0 {
		return didURL[:i]
	}
	return didURL
}

// NewMessageID returns a fresh DIDComm message id (a UUID URN), independent of
// the Trust Task document's own id (binding §2).
func NewMessageID() string {
	return "urn:uuid:" + uuidV4()
}

func uuidV4() string {
	var b [16]byte
	_, _ = rand.Read(b[:])
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	const hexd = "0123456789abcdef"
	buf := make([]byte, 36)
	for i, p := range []int{0, 2, 4, 6, 9, 11, 14, 16, 19, 21, 24, 26, 28, 30, 32, 34} {
		buf[p] = hexd[b[i]>>4]
		buf[p+1] = hexd[b[i]&0x0f]
	}
	buf[8], buf[13], buf[18], buf[23] = '-', '-', '-', '-'
	return string(buf)
}
